src/include/storage/copier/rel_copier.h

```diff
@@ -0,0 +1,278 @@
+                                                                            std::move(filePath)} {};
+    virtual ~RelCopyMorsel() = default;
+
+public:
+    common::block_idx_t blockIdx;
+    common::offset_t tupleIdx;
+    uint64_t numTuples;
+    std::string filePath;
+};
+
+class CSVRelCopyMorsel : public RelCopyMorsel {
+public:
+    CSVRelCopyMorsel(std::string filePath, common::offset_t tupleIdx,
+        std::shared_ptr<arrow::RecordBatch> recordBatch)
+        : RelCopyMorsel{BLOCK_IDX_INVALID, tupleIdx, UINT64_MAX, std::move(filePath)},
+          recordBatch{std::move(recordBatch)} {}
+
+    std::shared_ptr<arrow::RecordBatch> recordBatch;
+};
+
+class RelCopySharedState {
+public:
+    RelCopySharedState(std::vector<std::string> filePaths,
+        std::unordered_map<std::string, FileBlockInfo> fileBlockInfos)
+        : filePaths{std::move(filePaths)}, fileIdx{0}, numTuples{0},
+          fileBlockInfos{std::move(fileBlockInfos)}, blockIdx{0} {};
+    virtual ~RelCopySharedState() = default;
+
+    virtual std::unique_ptr<RelCopyMorsel> getMorsel();
+
+public:
+    std::vector<std::string> filePaths;
+    common::vector_idx_t fileIdx;
+    common::offset_t numTuples;
+
+protected:
+    std::unordered_map<std::string, FileBlockInfo> fileBlockInfos;
+    common::block_idx_t blockIdx;
+    std::mutex mtx;
+};
+
+class CSVRelCopySharedState : public RelCopySharedState {
+public:
+    CSVRelCopySharedState(std::vector<std::string> filePaths,
+        std::unordered_map<std::string, FileBlockInfo> fileBlockInfos,
+        common::CSVReaderConfig* csvReaderConfig, catalog::TableSchema* tableSchema)
+        : RelCopySharedState{std::move(filePaths), std::move(fileBlockInfos)},
+          csvReaderConfig{csvReaderConfig}, tableSchema{tableSchema} {};
+
+    std::unique_ptr<RelCopyMorsel> getMorsel() final;
+
+private:
+    common::CSVReaderConfig* csvReaderConfig;
+    catalog::TableSchema* tableSchema;
+    std::shared_ptr<arrow::csv::StreamingReader> reader;
+};
+
+class RelCopier {
+public:
+    RelCopier(std::shared_ptr<RelCopySharedState> sharedState,
+        const common::CopyDescription& copyDesc, catalog::RelTableSchema* schema,
+        DirectedInMemRelData* fwdRelData, DirectedInMemRelData* bwdRelData,
+        std::vector<common::LogicalType> pkColumnTypes, std::vector<PrimaryKeyIndex*> pkIndexes)
+        : sharedState{std::move(sharedState)}, copyDesc{copyDesc}, schema{schema},
+          fwdRelData{fwdRelData}, bwdRelData{bwdRelData},
+          pkColumnTypes{std::move(pkColumnTypes)}, pkIndexes{std::move(pkIndexes)} {}
+    virtual ~RelCopier() = default;
+
+    void execute(processor::ExecutionContext* executionContext);
+
+    virtual std::unique_ptr<RelCopier> clone() const {
+        return std::make_unique<RelCopier>(
+            sharedState, copyDesc, schema, fwdRelData, bwdRelData, pkColumnTypes, pkIndexes);
+    }
+    virtual void finalize() {
```

Comment:
I think you can make relCopier a pure virtual class by making this function pure virtual